arch/x86/mp.c

```diff
@@ -87,18 +88,29 @@ status_t arch_mp_send_ipi(mp_cpu_mask_t target, mp_ipi_t ipi) {
 void arch_mp_init_percpu(void) {}
 
 uint32_t x86_get_apic_id_from_hardware(void) {
-    // read the apic id out of cpuid leaf 1, which should be present if SMP is enabled.
-    uint32_t apic_id, unused;
-    cpuid(0x1, &unused, &apic_id, &unused, &unused);
-
-    apic_id >>= 24;
-
-    // TODO: read full 32bit apic id from x2apic msr if available
-
-    return apic_id;
+    uint32_t unused = 0, ecx;
+    cpuid(0x1, &unused, &unused, &ecx, &unused);
+
+    if (ecx & (1u << 21)) {
+        // read full 32bit apic id from x2apic msr if available
+        return (uint32_t)read_msr(0x802);
```

Comment:
I like the idea of this. I think it might be a bit cleaner if it made a call into the lapic.c code which also is responsible for bringing up the x2apic per cpu to request the apic id there. This way it can initialize the x2apic without needing to do it again here.

I think basically moving this code into the lapic.c file would suffice, and have it automatically enable x2apic mode if it's called would work.